core/src/main/java/org/elasticsearch/common/geo/builders/GeometryCollectionBuilder.java

```diff
@@ -132,4 +130,39 @@ public Shape build() {
         //note: ShapeCollection is probably faster than a Multi* geom.
     }
 
+    @Override
+    public int hashCode() {
+        return Objects.hash(shapes);
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (this == obj) {
+            return true;
+        }
+        if (obj == null || getClass() != obj.getClass()) {
+            return false;
+        }
+        GeometryCollectionBuilder other = (GeometryCollectionBuilder) obj;
+        return Objects.equals(shapes, other.shapes);
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeVInt(shapes.size());
+        for (ShapeBuilder shape : shapes) {
+            out.writeShape(shape);
+        }
+    }
+
+    @Override
+    public GeometryCollectionBuilder readFrom(StreamInput in) throws IOException {
+        GeometryCollectionBuilder geometryCollectionBuilder = new GeometryCollectionBuilder();
+        int shapes = in.readVInt();
```

Comment:
I added a test to GeoShapeQueryTests that tests that empty geometry collection doesn't cause an error but returns zero results. So I think this is a valid edge case, or do you think we should prevent this in the parser and then also during serialization? In this case I'd open a follow up PR for this.